### PR TITLE
gh-428: Add GitHub issue to description of PR

### LINKED ISSUE #428: Add link to the GitHub Issue from the PR
We used to automatically append a link to the bpo ticket from the GitHub PR, by looking at the PR title's issue number.

In https://github.com/python/bedevere/pull/420, the functionality was dropped. I had thought that GitHub UI would automatically link the issue number when referenced like this, however this is not a case. GitHub would automatically link the issue in the PR descriptions or in the PR comments, but not in the title.

To make it easier for us to navigate to the issue from the title, we should add back the reference to the issue in the PR description.
Upon seeing the `GH-` issue number in the PR title, bedevere should edit the description and add the issue number there.
We don't need to add it as a link, because GitHub can automatically render `gh-` as links, so just adding the text `gh-NNNN` at the end would be sufficient.

This text should be added only if the issue number has not already mentioned, to avoid duplicated links.

### PR DESCRIPTION
**Pull request by ryanlong1004**
Closes #428

### CODE REVIEW

File: bedevere/util.py [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
 
+
+DEFAULT_BODY=""
+TAG_NAME = "issue-number"
 NEWS_NEXT_DIR = "Misc/NEWS.d/next/"
+CLOSING_TAG = f"<!-- /{TAG_NAME} -->"
+BODY = f"""\
+{{body}}
+<!-- {TAG_NAME}: gh-{{issue_number}} -->
+https://bugs.python.org/issue{{issue_number}}
```
Comment (Mariatta) [inline bedevere/util.py:15]: It should just be `gh-{{issue_number}}`, no links. GitHub will automatically link that in the UI.
We shouldn't have links to b.p.o anymore.

**Review by Mariatta — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 2]
```diff
@@ -1,14 +1,42 @@
 aiohttp==3.8.1
+aiosignal==1.2.0
```
[the remark this answers is not in the document]
Comment (ryanlong1004) [reply requirements.txt:2]: I had to install these in order to run pytest.

[the remark this answers is not in the document]
Comment (Mariatta) [reply requirements.txt:2]: Can you try using the `dev-requirements.txt` instead?

File: bedevere/util.py [hunk 3]
```diff
@@ -76,6 +87,15 @@ async def issue_for_PR(gh, pull_request):
     """Get the issue data for a pull request."""
     return await gh.getitem(pull_request["issue_url"])
 
+async def patch_body(gh, event, issue_number):
+    """Updates the description of a PR with the gh issue number if it exists.
+
+    Does nothing if the 'body' key already exists on the PR.
+    """
+    if "body" in event.data["pull_request"]:
```
[the remark this answers is not in the document]
Comment (ryanlong1004) [reply bedevere/util.py]: Fixed in updated PR.

File: tests/test_gh_issue.py [hunk 4]
```diff
@@ -28,6 +30,13 @@ async def post(self, url, *, data):
         self.post_data.append(data)
         return self._post_return
 
+    async def patch(self, url, *, data):
+        self.patch_url.append(url)
+        self.patch_data.append(data)
+        return self._post_return
```
[the remark this answers is not in the document]
Comment (ryanlong1004) [reply tests/test_gh_issue.py]: Fixed in updated PR.

File: bedevere/util.py (hunk 1 above)
Comment (ryanlong1004) [reply bedevere/util.py:14]: Fixed in updated PR.

**Comment by hugovk** [comment]
@ryanlong1004 Please can you resolve the conflicts?

File: requirements.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanlong1004) [reply requirements.txt]: good to go

File: bedevere/gh_issue.py [hunk 5]
```diff
@@ -46,6 +46,7 @@ async def set_status(
         )
         if issue_found:
             status = create_success_status(issue_number, kind=issue_kind)
+
```
Comment (JelleZijlstra) [inline bedevere/gh_issue.py:49]: Undo the unnecessary change here
```suggestion
```

File: bedevere/util.py [hunk 6]
```diff
@@ -76,6 +86,15 @@ async def issue_for_PR(gh, pull_request):
     """Get the issue data for a pull request."""
     return await gh.getitem(pull_request["issue_url"])
 
+async def patch_body(gh, pull_request, issue_number):
+    """Updates the description of a PR with the gh issue number if it exists.
+
+    returns if body exists with issue_number
+    """
+    if "body" in pull_request and pull_request["body"] is not None and issue_number in pull_request["body"]:
```
Comment (JelleZijlstra) [inline bedevere/util.py:94]: A simple "in" check could have false positives. Should probably check for `f"GH-{issue_number}\n"` or similar.

File: tests/test_gh_issue.py [hunk 7]
```diff
@@ -28,6 +31,13 @@ async def post(self, url, *, data):
         self.post_data.append(data)
         return self._post_return
 
+    async def patch(self, url, *, data):
+        self.patch_url.append(url)
+        self.patch_data.append(data)
+        return self._patch_return
+
+
+
```
Comment (JelleZijlstra) [inline tests/test_gh_issue.py:40]:
```suggestion

```
Too many blank lines

File: bedevere/util.py (hunk 6 above)
Comment (ryanlong1004) [reply bedevere/util.py:93]: Fixed.

File: bedevere/util.py [hunk 8]
```diff
@@ -76,6 +85,15 @@ async def issue_for_PR(gh, pull_request):
...
+        return
+    await gh.patch(pull_request["url"], data=BODY.format(body=DEFAULT_BODY, issue_number=issue_number))
```
Comment (JelleZijlstra) [inline bedevere/util.py:95]: Seems like this drops the existing body.

Comment (ryanlong1004) [reply bedevere/util.py]: Updated.

File: bedevere/util.py [hunk 9]
```diff
@@ -3,7 +3,16 @@
...
 
+
+DEFAULT_BODY = ""
+TAG_NAME = "issue-number"
 NEWS_NEXT_DIR = "Misc/NEWS.d/next/"
+CLOSING_TAG = f"<!-- /{TAG_NAME} -->"
+BODY = f"""\
+{{body}}
+gh-{{issue_number}}
+{CLOSING_TAG}
```
Comment (JelleZijlstra) [inline bedevere/util.py:14]: Why is there a closing tag but no opening tag?

I'm not sure what the tag is for, but my best guess is that it's so the bot can reliably recognize that it has already added a link. So maybe we should also add an open tag and use that to check whether to add the issue number.

File: tests/test_gh_issue.py [hunk 10]
```diff
@@ -28,6 +31,10 @@ async def post(self, url, *, data):
         self.post_data.append(data)
         return self._post_return
 
+    async def patch(self, url, *, data):
+        self.patch_url.append(url)
+        self.patch_data.append(data)
+        return self._patch_return
```
Comment (JelleZijlstra) [inline tests/test_gh_issue.py:37]: Seems like this doesn't get used.

Comment (ryanlong1004) [reply tests/test_gh_issue.py]: removed.

File: bedevere/util.py (hunk 9 above)
Comment (ryanlong1004) [reply bedevere/util.py:14]: Taken from [2aee86a31df388b9292a28afb1c28a473c4d577b] (https://github.com/python/bedevere/blob/2aee86a31df388b9292a28afb1c28a473c4d577b/bedevere/bpo.py) there was no opening tag their either.

**Comment by Mariatta** [comment]
Thanks for your work @ryanlong1004!

Would you be able to increase the coverage to 100%?

File: bedevere/util.py (hunk 9 above)
Comment (erlend-aasland) [reply bedevere/util.py:14]: @JelleZijlstra, can this conversation be resolved?

**Comment by erlend-aasland** [comment]
@Mariatta, you requested changes. Are you happy with the PR as it stands? Please mark it approved, if so.

File: bedevere/util.py (hunk 9 above)
Comment (JelleZijlstra) [reply bedevere/util.py:14]: Though it seems we're not using the tag now. If so, we should remove the closing tag too.

File: tests/test_util.py [hunk 11]
```diff
@@ -157,3 +158,36 @@ async def test_get_pr_for_commit_not_found():
...
+        await util.patch_body(gh, vals, "1234")
+        mock.assert_not_called()
+    with patch.object(gh, "patch") as mock:
+        vals["body"] = None
+        await util.patch_body(gh, vals, "1234")
+        mock.assert_called_once()
+    with patch.object(gh, "patch") as mock:
+        vals["body"] = ""
+        await util.patch_body(gh, vals, "1234")
+        data = "\ngh-1234\n<!-- /issue-number -->\n"
```
Comment (ezio-melotti) [inline tests/test_util.py:191]: This should fix the test too (untested):
```suggestion
        data = "<!-- gh-issue-number: gh-1234 -->\n* gh-1234\n<!-- /gh-issue-number -->\n"
```

File: bedevere/util.py [hunk 12]
```diff
@@ -3,7 +3,18 @@
...
 
+
+DEFAULT_BODY = ""
+TAG_NAME = "gh-issue-number"
 NEWS_NEXT_DIR = "Misc/NEWS.d/next/"
+CLOSING_TAG = f"<!-- /{TAG_NAME} -->"
+BODY = f"""\
+{{body}}
+
+<!-- {TAG_NAME}: gh-{{issue_number}} -->
```
Comment (ezio-melotti) [inline bedevere/util.py:14]:
```suggestion
<!-- {TAG_NAME}: gh-{{issue_number}} -->
```

File: tests/test_util.py [hunk 13]
```diff
@@ -157,3 +158,36 @@ async def test_get_pr_for_commit_not_found():
...
+        await util.patch_body(gh, vals, "1234")
+        mock.assert_not_called()
+    with patch.object(gh, "patch") as mock:
+        vals["body"] = None
+        await util.patch_body(gh, vals, "1234")
+        mock.assert_called_once()
+    with patch.object(gh, "patch") as mock:
+        vals["body"] = ""
+        await util.patch_body(gh, vals, "1234")
+        data = "<!-- gh-issue-number: gh-1234 -->\n* gh-1234\n<!-- /gh-issue-number -->\n"
```
Comment (ezio-melotti) [inline tests/test_util.py:191]:
```suggestion
        data = "\n\n<!-- gh-issue-number: gh-1234 -->\n* gh-1234\n<!-- /gh-issue-number -->\n"
```
